tests/unit/h/security/predicates_test.py

```diff
@@ -168,17 +174,156 @@ def test_group_created_by_user(self, identity, group_context, factories):
+                # contexts (e.g. the WebSocket) there can be detached objects
+                # that won't compare equal to other objects with the same id.
+                user=factories.User.build(id=authenticated_user.id),
+                roles=[role],
+            )
+        )
+
+        assert (
+            predicates.group_has_user_as_owner(identity, group_context)
+            == expected_result
+        )
+
+    def test_group_has_user_as_owner_when_user_isnt_member(
+        self, identity, group_context
+    ):
+        assert predicates.group_has_user_as_owner(identity, group_context) is False
+
+    @pytest.mark.parametrize(
+        "role,expected_result",
+        [
+            (GroupMembershipRoles.MEMBER, False),
+            (GroupMembershipRoles.MODERATOR, False),
+            (GroupMembershipRoles.ADMIN, True),
+            (GroupMembershipRoles.OWNER, False),
+        ],
+    )
+    def test_group_has_user_as_admin_when_user_is_member(
+        self,
+        authenticated_user,
+        identity,
+        group_context,
+        role,
+        expected_result,
+        factories,
+    ):
+        group_context.group.memberships.append(
+            GroupMembership(
+                # Rather than passing `user=user` we create another user object
+                # with the same id. This is to ensure that the predicate tests
+                # based on id rather than on object equality because in some
+                # contexts (e.g. the WebSocket) there can be detached objects
+                # that won't compare equal to other objects with the same id.
+                user=factories.User.build(id=authenticated_user.id),
+                roles=[role],
+            )
+        )
+
+        assert (
+            predicates.group_has_user_as_admin(identity, group_context)
+            == expected_result
+        )
+
+    def test_group_has_user_as_admin_when_user_isnt_member(
+        self, identity, group_context
+    ):
+        assert predicates.group_has_user_as_admin(identity, group_context) is False
+
+    @pytest.mark.parametrize(
+        "role,expected_result",
+        [
+            (GroupMembershipRoles.MEMBER, False),
+            (GroupMembershipRoles.MODERATOR, True),
+            (GroupMembershipRoles.ADMIN, False),
+            (GroupMembershipRoles.OWNER, False),
+        ],
+    )
+    def test_group_has_user_as_moderator_when_user_is_member(
+        self,
+        authenticated_user,
+        identity,
+        group_context,
+        role,
+        expected_result,
+        factories,
     ):
-        # Note we don't use the same literal objects here. It's important
-        # that we test based on the values not Python equality as the objects
-        # are detached in the WS and don't evaluate as equal
-        identity.user.groups = [factories.Group.build(id=i) for i in range(3)]
-        group_context.group = factories.Group.build(id=1 if matching else 100)
```

Comment:
This test was written in a weird way that was coupled to the way the predicate was implemented (based on `user.groups` instead of `group.members`). Rewrote the test to work the same as the new tests that I added for the other predicates.